jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java

```diff
@@ -59,17 +78,17 @@ public Content.Chunk read()
 
         try
         {
-            byte[] buffer = new byte[getBufferSize()];
-            int read = inputStream.read(buffer);
+            ByteBuffer buffer = bufferPool.acquire(getBufferSize(), false);
+            int read = inputStream.read(buffer.array());
             if (read < 0)
             {
                 close();
                 return Content.Chunk.EOF;
             }
             else
             {
-                ByteBuffer byteBuffer = ByteBuffer.wrap(buffer, 0, read);
-                return Content.Chunk.from(byteBuffer, false);
+                buffer.position(read);
```

Comment:
Buffer should be flipped and sliced.

Reply:
shy is slice needed?  The buffer is not retained nor reused by this class, so once the consumer calls release it is put back into the pool.

Perhaps it could be held expecting a release?  But this is simpler for now unless there is a performance indication that we need to do better.

Reply:
Hmmm ContentSourceTest tests all the failure modes, but does test actually reading the content!  So this code is a) wrong; b) not tested!  fixing....